README.md
Виджет комментарии VK
===================================

Информация
-------------------

Виджет комментарии VK

Пример использования

```php

<?= \skeeks\cms\vk\comments\VkCommentsWidget::widget([
    'namespace' => 'VkCommentsWidget-main',
    'apiId'     => 4982033
]); ?>

```


> [![skeeks!](https://gravatar.com/userimage/74431132/13d04d83218593564422770b616e5622.jpg)](http://skeeks.com)  
<i>SkeekS CMS (Yii2) — быстро, просто, эффективно!</i>  
[skeeks.com](http://skeeks.com) | [cms.skeeks.com](http://cms.skeeks.com) | [marketplace.cms.skeeks.com](http://marketplace.cms.skeeks.com)


CHANGELOG.md
Процесс разработки модуля.
==============
  
1.3.0
-----------------
  * Адаптация под SkeekS CMS 5
  
1.2.0
-----------------
  * Адаптация под SkeekS CMS 2.5

1.1.0
-----------------
  * Добавлена адаптивность

1.0.0
-----------------
  * Стабильный релиз


src/views/default.php
<?php
/**
 * @author Semenov Alexander <semenov@skeeks.com>
 * @link http://skeeks.com/
 * @copyright 2010 SkeekS (СкикС)
 * @date 03.07.2015
 */
/* @var $this yii\web\View */
/* @var $widget \skeeks\cms\vk\comments\VkCommentsWidget */
?>

<? if ($widget->apiId) : ?>
    <?
    $this->registerJs(<<<JS
        VK.init({apiId: {$widget->apiId}, onlyWidgets: true});


        (function(sx, $, _)
        {
            sx.classes.VkWidgetComments = sx.classes.Component.extend({

                getResutlOptions: function()
                {
                    var options = this.get('options');
                    var jQWigetWrapper = $("#" + this.get('id'));
                    var jQWigetWrapperParent = jQWigetWrapper.parent();

                    if (this.get('adaptiveWith'))
                    {
                        if (jQWigetWrapperParent.length)
                        {
                            options = _.extend(options, {
                                'width' : jQWigetWrapperParent.width()
                            });
                        }
                    }

                    return options;
                },

                buildWidget: function()
                {
                    VK.Widgets.Comments(this.get('id'), this.getResutlOptions());

                    return this;
                },

                destroyWidget: function()
                {
                    $("#" + this.get('id')).empty();
                    return this;
                },

                _onDomReady: function()
                {
                    var self = this;
                    $( window ).resize(function() {
                       self.destroyWidget().buildWidget();
                    });

                    this.buildWidget();
                }

            });
        })(sx, sx.$, sx._);

        new sx.classes.VkWidgetComments({
            'adaptiveWith' : {$widget->adaptiveWith},
            'id' : '{$widget->id}',
            'options' : {$widget->getJsonOptions()},
        });


JS
    );
    ?>
    <div class="sx-vk-widget-comments-wrapper">
        <div id="<?= $widget->id; ?>"></div>
    </div>
<? endif; ?>


views/_settingsForm.php
<?php
/**
 * @author Semenov Alexander <semenov@skeeks.com>
 * @link http://skeeks.com/
 * @copyright 2010 SkeekS (СкикС)
 * @date 27.05.2015
 */
/* @var $this yii\web\View */
?>
<?= $form->fieldSet('Отображение'); ?>
    <?= $form->field($model, 'viewFile')->textInput(); ?>
<?= $form->fieldSetEnd(); ?>

<?= $form->fieldSet('Параметр виджета'); ?>
    <?= $form->fieldInputInt($model, 'apiId')->hint('https://vk.com/apps?act=manage - посмотреть тут');; ?>
    <?= $form->fieldInputInt($model, 'limit')->hint('количество комментариев на странице (целое число 5-100)'); ?>
    <?= $form->fieldInputInt($model, 'width')->hint('целое число > 300'); ?>
    <?= $form->fieldInputInt($model, 'height')->hint('задает максимальную высоту виджета в пикселях. Целое число > 500. Если равно 0, то высота не ограничена. Если содержимое виджета больше, чем максимально допустимое, то появляется внутренняя прокрутка. Значение по умолчанию - 0.'); ?>
    <?= $form->field($model, 'attach')->checkboxList(\skeeks\cms\vk\comments\VkCommentsWidget::$possibleAttach)->hint('задает возможность создания прикреплений к комментариям. Строка, содержащая перечисленные через запятую типы допустимых прикреплений либо false в случае отключения этой функции. Возможные типы: graffiti, photo, audio, video, link.'); ?>
    <?= $form->field($model, 'autoPublish')->radioList([
        '1' => 'включена',
        '0' => 'выключена',
    ])->hint('автоматическая публикация комментария в статус пользователю (0 - выключена, 1 - включена). Значение по умолчанию - 1.'); ?>
    <?= $form->field($model, 'norealtime')->radioList([
        '1' => 'отключено',
        '0' => 'включено',
    ])->hint('отключает обновление ленты комментариев в реальном времени. (1 - отключено, 0 - включено). Значение по умолчанию - 0.'); ?>
    <?= $form->field($model, 'mini')->radioList([
        '1' => 'включено',
        '0' => 'выключено',
        'auto' => 'выбирать автоматически в зависимости от доступной ширины',
    ])->hint('использовать ли минималистичный вид виджета - уменьшенный шрифт, уменьшенные миниатюры прикреплений, уменьшенные профильные изображения для комментариев 2го уровня. (1 - включено, 0 - выключено, \'auto\' - выбирать автоматически в зависимости от доступной ширины). Значение по умолчанию - auto'); ?>

<?= $form->fieldSetEnd(); ?>



src/VkCommentsWidget.php
<?php
/**
 * @author Semenov Alexander <semenov@skeeks.com>
 * @link http://skeeks.com/
 * @copyright 2010 SkeekS (СкикС)
 * @date 03.07.2015
 */
namespace skeeks\cms\vk\comments;

use skeeks\cms\base\WidgetRenderable;
use skeeks\cms\components\Cms;
use yii\helpers\ArrayHelper;
use yii\helpers\Json;
use yii\widgets\ActiveForm;

/**
 * @property string $jsonOptions
 *
 * Class VkCommentsWidget
 * @package skeeks\cms\vk\comments
 */
class VkCommentsWidget extends WidgetRenderable
{
    const ATTACH_PHOTO      = 'photo';
    const ATTACH_GRAFFITI   = 'graffiti';
    const ATTACH_AUDIO      = 'audio';
    const ATTACH_VIDEO      = 'video';
    const ATTACH_LINK       = 'link';

    public $adaptiveWith    = true;
    public $limit           = 10;
    public $width           = 500;
    public $autoPublish     = 1;
    public $mini            = 'auto';
    public $height          = 0;
    public $norealtime      = 0;

    public $attach = [
        self::ATTACH_GRAFFITI,
        self::ATTACH_PHOTO,
        self::ATTACH_AUDIO,
        self::ATTACH_VIDEO,
        self::ATTACH_LINK,
    ];

    static public $possibleAttach = [
        self::ATTACH_GRAFFITI   => 'Граффити',
        self::ATTACH_PHOTO      => 'Фотографии',
        self::ATTACH_VIDEO      => 'Видео',
        self::ATTACH_AUDIO      => 'Аудио',
        self::ATTACH_LINK       => 'Ссылки',
    ];

    public $apiId;

    static public function descriptorConfig()
    {
        return array_merge(parent::descriptorConfig(), [
            'name' => 'Виджет комментарии VK'
        ]);
    }


    public function renderConfigForm(ActiveForm $form)
    {
        echo \Yii::$app->view->renderFile(__DIR__ . '/views/_settingsForm.php', [
            'form'  => $form,
            'model' => $this
        ], $this);
    }

    public function attributeLabels()
    {
        return array_merge(parent::attributeLabels(),
        [
            'apiId'                 => 'ID приложения vk',
            'limit'                 => 'Количество записей',
            'attach'                => 'Возможность создания прикреплений к комментариям.',
            'autoPublish'           => 'Автоматическая публикация комментария в статус пользователю',
            'mini'                  => 'Минималистичный вид виджета',
            'height'                => 'Максимальную высоту виджета в пикселях',
            'width'                 => 'Ширина блока',
            'norealtime'            => 'Отключить обновление ленты комментариев в реальном времени.',
        ]);
    }

    public function rules()
    {
        return ArrayHelper::merge(parent::rules(),
        [
            [['apiId'], 'integer'],
            [['apiId'], 'required'],
            [['limit'], 'integer', 'min' => 5, 'max' => 100],
            [['width'], 'integer', 'min' => 300],
            [['attach'], 'safe'],
            [['autoPublish'], 'in', 'range' => [0,1]],
            [['mini'], 'in', 'range' => [0,1,'auto']],
            [['height'], 'integer'],
            [['norealtime'], 'in', 'range' => [0,1]],
        ]);
    }

    /**
     * @return string
     */
    public function getJsonOptions()
    {
        return Json::encode([
            'width' => (int) $this->width,
            'limit' => (int) $this->limit,
            'attach' => implode(',', (array) $this->attach),
            'autoPublish' => (int) $this->autoPublish,
            'mini' => $this->mini,
            'height' => (int) $this->height,
            'norealtime' => (int) $this->norealtime,
        ]);
    }
    /**
     * @return string
     */
    public function run()
    {
        VkAsset::register($this->view);
        return parent::run();
    }
}
